File: leetcode/js/climbing-stairs.js
/**
 * @param {number} n
 * @return {number}
 */
var climbStairs = function (n) {
  const cache = {};
  const climbing = (n) => {
    if (n <= 3) {
      return n;
    }

    if (!(n in cache)) {
      cache[n] = climbing(n - 1) + climbing(n - 2);
    }

    return cache[n];
  };

  return climbing(n);
};

var climbStairs = function (n) {
  const cache = Array(n + 1).fill(1);

  for (let i = 2; i <= n; i++) {
    cache[i] = cache[i - 1] + cache[i - 2];
  }

  return cache[n];
};

var climbStairs = function (n) {
  let a = 0;
  let b = 1;
  let result = 0;

  for (let i = 1; i <= n; i++) {
    result = a + b;
    a = b;
    b = result;
  }

  return result;
};

File: leetcode/rust/climbing-stairs.rs
impl Solution1 {
    pub fn climb_stairs(n: i32) -> i32 {
        let n = n as usize;
        let mut cache = vec![1; n + 1];

        for i in 2..=n {
            cache[i] = cache[i - 1] + cache[i - 2];
        }

        cache[n]
    }
}

impl Solution2 {
    pub fn climb_stairs(n: i32) -> i32 {
        let mut a = 0;
        let mut b = 1;
        let mut result = 0;

        for i in 1..=n {
            result = a + b;
            a = b;
            b = result;
        }

        result
    }
}

impl Solution3 {
    pub fn climb_stairs(n: i32) -> i32 {
        fn climbing(n: usize, dp: &mut Vec<i32>) -> i32 {
            if n <= 1 {
                return 1;
            }

            if dp[n] != 0 {
                return dp[n];
            }

            dp[n] = climbing(n - 1, dp) + climbing(n - 2, dp);
            dp[n]
        }

        let n = n as usize;
        let mut dp = vec![0; n + 1];

        climbing(n, &mut dp)
    }
}
